docs/intro.md

```diff
@@ -0,0 +1,1014 @@
+
+```
+HSET person:3 name "Summer Smith" age 17
+```
+
+A registered function is by definition nonblocking, so any results it returns can only be obtained with the `RG.GETRESULTS` command and by specifying its respective execution ID. At this point we know there is just one registered function and that it had executed just once, so we can use `RG.DUMPEXECUTIONS` output to get the last execution's ID and then read the results:
+
+!!! example "Example: dumping executions"
+    ```
+    127.0.0.1:6379> HSET person:3 name "Summer Smith" age 17
+    (integer) 2
+    127.0.0.1:6379> RG.DUMPEXECUTIONS
+    1) 1) "executionId"
+       2) "0000000000000000000000000000000000000000-1"
+       3) "status"
+       4) "done"
+    2) 1) "executionId"
+       2) "0000000000000000000000000000000000000000-0"
+       3) "status"
+       4) "done"
+    127.0.0.1:6379> RG.GETRESULTS 0000000000000000000000000000000000000000-1
+    1) 1) (integer) 17
+    2) (empty array)
+    ```
+
+Note that the executions' list now consists of two entries: the first is the most recent one generated by the registered function, and the second entry is from our previous nonblocking execution of the batch function.
+
+The result "17" is technically correct, in the sense that it is the maximum of inputs and in this case, a single record. To have an event-driven maximum, we'll store its current value in the database.
+
+## Writing Data
+The RedisGears Python API ships with the [`execute()` function](runtime.md#execute), which allows the execution of arbitrary Redis commands in the database. RedisGears functions can call `execute()` for accessing the data during their flow, both for reading and writing, allowing the enrichment of inputs and persistence of results.
+
+We'll complete the implementation that seeks an event-driven maximum by storing the current maximum value in another Redis key called `age:maximum`. We'll also have two RedisGears functions - one for handling events and another for batch processing the initial data:
+
+```python
+pattern = 'person:*'  # input keys' pattern
+
+def age(x):
+  ''' Extracts the age from a person's record '''
+  return int(x['value']['age'])
+
+def maximum(a, x):
+  ''' Returns the maximum '''
+  a = a if a else 0  # initialize the accumulator
+  if x > a:          # if a new maximum was found
+    a = x            # set accumulator to maximum
+  return a           # finally output accumulator
+
+def cas(x):
+  ''' Checks and sets the current maximum '''
+  k = 'age:maximum'
+  v = execute('GET', k)   # read key's current value
+  v = v if v else 0       # initialize to 0 if None
+  if x > val:             # if a new maximum found
+    execute('SET', k, x)  # set key to new value
+
+# Event handling function registration
+GB() \
+  .map(age) \
+  .foreach(cas) \
+  .register(pattern)
+
+# Batch processing function execution
+GB() \
+  .map(age) \
+  .accumulate(maximum) \
+  .foreach(cas) \
+  .run(pattern)
+
+## Expected result: [70]
```

Comment:
There is a concurrency issue with this example in which the run might override a data that was set by the register which is the correct max value. The correct way to solve it is having the run to be sync but its not supported currently with run ...

Reply:
Hmm... this should be perhaps mentioned, or maybe fix the example to something else?

Reply:
Possible example fix (that would also highlight the current concurrency issue) is to add a version field. Alternatively, I can make it into an avg example... a hash w/ sum and count would not suffer from concurrency as these would only be incr'd

Reply:
With hash of sum and count you might go over the same key twice ..